osu.Game/Screens/Edit/Compose/Components/LinedPositionSnapGrid.cs

```diff
@@ -0,0 +1,165 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Shapes;
+using osu.Framework.Utils;
+using osuTK;
+
+namespace osu.Game.Screens.Edit.Compose.Components
+{
+    public abstract partial class LinedPositionSnapGrid : PositionSnapGrid
+    {
+        protected void GenerateGridLines(Vector2 step, Vector2 drawSize)
+        {
+            if (Precision.AlmostEquals(step, Vector2.Zero))
+                return;
+
+            int index = 0;
+
+            // Make lines the same width independent of display resolution.
+            float lineWidth = DrawWidth / ScreenSpaceDrawQuad.Width;
+            float rotation = MathHelper.RadiansToDegrees(MathF.Atan2(step.Y, step.X));
+
+            List<Box> generatedLines = new List<Box>();
+
+            while (true)
+            {
+                Vector2 currentPosition = StartPosition.Value + index++ * step;
+
+                if (!lineDefinitelyIntersectsBox(currentPosition, step.PerpendicularLeft, drawSize, out var p1, out var p2))
+                {
+                    if (!isMovingTowardsBox(currentPosition, step, drawSize))
+                        break;
+
+                    continue;
+                }
+
+                var gridLine = new Box
+                {
+                    Colour = Colour4.White,
+                    Alpha = 0.1f,
+                    Origin = Anchor.Centre,
+                    RelativeSizeAxes = Axes.None,
+                    Width = lineWidth,
+                    Height = Vector2.Distance(p1, p2),
+                    Position = (p1 + p2) / 2,
+                    Rotation = rotation,
+                };
+
+                generatedLines.Add(gridLine);
+            }
+
+            if (generatedLines.Count == 0)
+                return;
+
+            generatedLines.First().Alpha = 0.2f;
+
+            AddRangeInternal(generatedLines);
+        }
+
+        private bool isMovingTowardsBox(Vector2 currentPosition, Vector2 step, Vector2 box)
+        {
+            return (currentPosition + step).LengthSquared < currentPosition.LengthSquared ||
+                   (currentPosition + step - box).LengthSquared < (currentPosition - box).LengthSquared;
+        }
```

Comment:
How does this work? Why is checking distance against the top left and bottom right corner sufficient? I'd imagine that you'd wanna check all four corners (although I can't find quick counterexamples)

Reply:
I tried to come up with at least an intuition as to why this is enough but I'm not sure I convinced myself so I guess I'll just have to take this at face value and fix it if it ever breaks.